### PR TITLE
rgw: Add a "disable replication" flag to bucket notification configuration

### PR DESCRIPTION
**Pull request by DhruvSkyy**
For Enhancement #68104
https://tracker.ceph.com/issues/68104

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/rgw/rgw_rest_pubsub.cc [hunk 1]
```diff
@@ -396,16 +396,16 @@ class RGWPSCreateTopicOp : public RGWOp {
...
+//  if (!driver->is_meta_master()) {
+//    op_ret = rgw_forward_request_to_master(
+//        this, *s->penv.site, s->owner.id, &bl_post_body, nullptr, s->info, y);
+//    if (op_ret < 0) {
+//      ldpp_dout(this, 4)
+//          << "CreateTopic forward_request_to_master returned ret = " << op_ret
+//          << dendl;
+//      return;
+//    }
+// }
```
Comment (cbodley) [inline src/rgw/rgw_rest_pubsub.cc:408]: https://github.com/ceph/ceph/pull/59980 looks like a more complete version of this commit

Comment (DhruvSkyy) [reply src/rgw/rgw_rest_pubsub.cc:408]: Ah sorry, was meant to be a draft PR. Will commit a more complete fix to https://tracker.ceph.com/issues/68104 soon. 

Was thinking of adding another feature like how notification_v2 was added to disable multisite replication and checking for that, or adding a parameter to the topic creation HTTP request and checking for that?

**Comment by cbodley** [comment]
@yuvalif @kchheda3 can we please try to come up with a reasonable design for this use case? the multisite model expects metadata to be the same across all zones/zonegroups, and i don't think we should add special behavior for topics

**Comment by yuvalif** [comment]
@DhruvSkyy the usecase behind this fature request in a case where different zones need to send notifications to different brokers (which are local to the zone).
However, this would be specific to a topic, and not global configuration. There might be notifications that should be sent to topics which are not zone specific.
Therefore I would recomend implementing this as a new parameter to the topic, preventing it from being replicated.

**Comment by aza547** [comment]
> Therefore I would recomend implementing this as a new parameter to the topic, preventing it from being replicated.

I had in mind this would be a parameter on the notification rather than the topic. That would mean that topic metadata could still be synced multisite while the notification with "replicate=false" would not be.

The presence of the topic in all zones would not cause a problem as it would never have events added to a queue unless the notification was present on the zone. 

If this was present on the topic, would it not result in the scenario where there may be a notification on in a zone pointing to a topic that does not exist in that zone?

**Comment by yuvalif** [comment]
@DhruvSkyy and @aza547 few points came up in the refactoring meeting:
* as mentioned above, disabling per notifications is not possible due to the fact that these are bucket attributes, which are automatically synced as part of metadata sync
* disabling per topic may create issues and unexpected behaviour if the same topic is marked to be synced in one zone and not to be synced in another zone (same issue would happen even if we disable per notification)

given the above, the best proposal ( thank @cbodley and @aza547 ) would be to add a "ngetive" filter.
will capture this enhancements here:
* supprt negative filter: https://tracker.ceph.com/issues/68788. this would still require apps in different zones sending custom metadata with the zone information
* zone based filtering: https://tracker.ceph.com/issues/68789. with the above, it will make implementation on the application side easier

hence, closing this PR.

**Comment by 9401adarsh** [comment]
Have made a draft PR addressing the support for negative filters. PR Link: https://github.com/ceph/ceph/pull/61965

**Pull request closed without merging**